--- a/arch/arm/include/asm/memory.h
+++ b/arch/arm/include/asm/memory.h
@@ -185,6 +185,7 @@
 #define __phys_to_virt(x)	((x) - PHYS_OFFSET + PAGE_OFFSET)
 #endif
 #endif
+#endif /* __ASSEMBLY__ */
 
 #ifndef PHYS_OFFSET
 #ifdef PLAT_PHYS_OFFSET
@@ -193,6 +194,8 @@
 #define PHYS_OFFSET	UL(CONFIG_PHYS_OFFSET)
 #endif
 #endif
+
+#ifndef __ASSEMBLY__
 
 /*
  * PFNs are used to describe any physical page; this means
@@ -270,15 +273,17 @@
 #define virt_to_page(kaddr)	pfn_to_page(__pa(kaddr) >> PAGE_SHIFT)
 #define virt_addr_valid(kaddr)	((unsigned long)(kaddr) >= PAGE_OFFSET && (unsigned long)(kaddr) < (unsigned long)high_memory)
 
-/*
- * Optional coherency support.  Currently used only by selected
- * Intel XSC3-based systems.
- */
 #ifndef arch_is_coherent
-#define arch_is_coherent()		0
-#endif
-
-#endif
+#define arch_is_coherent()	0
+#endif
+
+#else
+#ifdef CONFIG_PLAT_BCM5301X
+#define __virt_to_phys(x)	((x) - PAGE_OFFSET + CONFIG_DRAM_BASE)
+#define __phys_to_virt(x)	((x) - CONFIG_DRAM_BASE + PAGE_OFFSET)
+#endif
+
+#endif	/* __ASSEMBLY__ */
 
 #include <asm-generic/memory_model.h>
 
