--- a/drivers/net/Makefile
+++ b/drivers/net/Makefile
@@ -69,4 +69,16 @@
 obj-$(CONFIG_USB_IPHETH)        += usb/
 obj-$(CONFIG_USB_CDC_PHONET)   += usb/
 
+#
+# Broadcom HND devices
+#
+obj-$(CONFIG_HND) += hnd/
+obj-$(CONFIG_BCM_CTF) += ctf/
+obj-$(CONFIG_DPSTA) += dpsta/
+obj-$(CONFIG_ET) += et/
+obj-$(CONFIG_BCM57XX) += bcm57xx/
+obj-$(CONFIG_WL) += wl/
+obj-$(CONFIG_EMF) += emf/ igs/
+obj-$(CONFIG_PROXYARP) += proxyarp/
+
 obj-$(CONFIG_HYPERV_NET) += hyperv/
