--- a/configure.ac
+++ b/configure.ac
@@ -4,36 +4,15 @@ AM_INIT_AUTOMAKE([-Wall -Werror foreign]
 AC_PROG_CC
 
 
-dnl -----------------------------------------------------------------
-dnl pcap check
-AC_MSG_CHECKING([whether to use libpcap for packet capture])
-
-AC_ARG_WITH([pcap],
-  AC_HELP_STRING( [--with-pcap@<:@=DIR@:>@],
-                  [use libpcap for packet capturing @<:@default=yes@:>@]),
-[
-	if test $withval = no
-	then
-		want_pcap=no
-	elif test $withval = yes
-	then
-		want_pcap=yes
-	else
-		want_pcap=yes
-		pcap_dir=$withval
-	fi
-],[
-	want_pcap=yes
-	pcap_dir=
-])
-if test "x$want_pcap" = "xno" ; then
-	AC_MSG_RESULT(no)
-else
-	AC_MSG_RESULT(yes)
-	AC_WIRESHARK_PCAP_CHECK
-fi
-dnl -----------------------------------------------------------------
-
+AC_ARG_WITH([pcap], [AC_HELP_STRING([ --with-pcap=TYPE], [How to link libpcap: dylib(Default) | dyload | stlib | PATH])],
+            [], [with_pcap="dylib"])
+AS_IF([test "x$with_pcap" = "xyes" || test -z "$with_pcap"], [with_pcap="dyload"])
+AS_IF([test "$with_pcap" = "dyload"], [AC_CHECK_LIB([dl], [dlsym])],
+      [AC_DEFINE(NO_DYLOAD, [], [Define NO_DYLOAD when not load libpcap dynamicly])
+       AS_CASE([$with_pcap],
+               [dylib], [AC_CHECK_LIB([pcap], [pcap_sendpacket])],
+               [stlib], [AC_SUBST(LDADD, libpcap.a)],
+               [AC_SUBST(LDADD, "$with_pcap")])])
 
 
 AC_CONFIG_HEADER([config.h])
