--- a/Makefile	2011-05-22 15:40:01.000000000 +0800
+++ b/Makefile	2012-03-13 13:53:04.953427000 +0800
@@ -1,14 +1,19 @@
-# Makefile under 2.6.25
-ifneq ($(KERNELRELEASE),)
-#kbuild syntax. dependency relationshsip of files and target modules are listed here.
-obj-m := napt66.o
+#
+# Makefile for napt66 kernel module
+#
+# Copyright (C) 2012 <tsl0922@sina.com>
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version
+# 2 of the License, or (at your option) any later version.
+#
+
+obj-$(CONFIG_NAPT66) := napt66.o
 napt66-objs := napt66_main.o napt66_conntrack.o napt66_nat.o napt66_hash_table.o napt66_ftp_alg.o
-else
-PWD  := $(shell pwd)
-KVER ?= $(shell uname -r)
-KDIR := /lib/modules/$(KVER)/build
-all:
-	$(MAKE) -C $(KDIR) M=$(PWD) modules
-clean:
-	rm -rf .*.cmd *.o *.mod.c *.ko .tmp_versions *.symvers *.order
+
+ifeq ($(MAKING_MODULES),1)
+export-objs := napt66.o
+
+-include $(TOPDIR)/Rules.make
 endif
